src/hooks/useTodos.ts

```diff
@@ -0,0 +1,212 @@
+
+  const todosFiltered = (() => {
+    switch (filterStatus) {
+      case FilterStatus.ACTIVE:
+        return todos.filter(todo => !todo.completed);
+      case FilterStatus.COMPLETED:
+        return todos.filter(todo => todo.completed);
+      default:
+        return todos;
+    }
+  })();
+
+  const addTodo = async (title: string) => {
+    const noSpaceQuery = title.trim();
+
+    if (!noSpaceQuery) {
+      showError(ToDoServiceErrors.Title);
+
+      return false;
+    }
+
+    const newTempTodo = {
+      id: 0,
+      title: noSpaceQuery,
+      completed: false,
+      userId: USER_ID,
+    };
+
+    setTempTodo(newTempTodo);
+    setLoadingTodo(0);
+
+    try {
+      const newTodo = await postTodo({
+        title: noSpaceQuery,
+        completed: false,
+        userId: USER_ID,
+      });
+
+      setTodos([...todos, newTodo]);
+
+      return true;
+    } catch {
+      showError(ToDoServiceErrors.UnableToAddTodo);
+
+      return false;
+    } finally {
+      setLoadingTodo(null);
+      setTempTodo(null);
+    }
+  };
+
+  const clearCompleted = async () => {
+    const todosDone = todos.filter(td => td.completed === true);
+
+    if (todosDone.length === 0) {
+      return;
+    }
+
+    const deleteResults = [];
+
+    for (const todo of todosDone) {
+      try {
+        await deleteTodo(todo.id);
+        deleteResults.push({ id: todo.id, success: true });
+      } catch (err) {
+        deleteResults.push({ id: todo.id, success: false });
+        showError(ToDoServiceErrors.UnableToDeleteTodo);
+      }
+    }
+
+    const deletedTodos = deleteResults
+      .filter(result => result.success)
+      .map(result => result.id);
+
+    const stayingTodos = todos.filter(todo => !deletedTodos.includes(todo.id));
+
+    setTodos(stayingTodos);
+  };
+
+  const toggleAll = async () => {
+    const newIfCompletedStatus = !allCompleted;
+
+    const todosDone = todos.map(td => ({
+      ...td,
+      completed: newIfCompletedStatus,
+    }));
+
+    try {
+      await Promise.all(
+        todos.map(todo =>
+          updateTodo(todo.id, { completed: newIfCompletedStatus }),
+        ),
+      );
+
+      setTodos(todosDone);
+    } catch (err) {
+      showError(ToDoServiceErrors.Unknown);
+    }
+  };
+
+  const clearQuery = () => setQuery('');
+
+  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+
+    addTodo(query).then(success => {
+      if (success) {
+        clearQuery();
+      }
+    });
+  };
+
+  return {
+    todos,
+    setTodos,
+    error,
+    setError,
+    isLoading,
+    setIsLoading,
+    filterStatus,
+    setFilterStatus,
+    todosFiltered,
+    tempTodo,
+    setTempTodo,
+    loadingTodo,
+    setLoadingTodo,
+    addTodo,
+    clearCompleted,
+    toggleAll,
+    showError,
+    query,
+    setQuery,
+    inputRef,
+    clearQuery,
+    handleSubmit,
+    allCompleted,
+    someCompleted,
+    activeCount,
+    completedCount,
+  };
```

Comment:
so big return and you passed all state in components
you are FE performance killer